### DISCUSSION TITLE
GdUnit3 v2.0.0-Beta Feedback

### DISCUSSION THREAD

**Discussion by MikeSchulze** [MikeSchulze/gdUnit3#228] [General]
# Hi GdUnit3 V2.0.0 Beta users,

You are invited to give me feedback on the GdUnit3 with C# beta support.

**Please note that C# test support is currently still in beta and not feature complete. If you have any problems or find any bugs, please report them.**

Please describe what you like and what you don't ;)
If you found bugs, please report them in the issue section of GitHib.
Also feel free to create feature requests.

Thanks in advance
Mike

**Comment by ZodmanPerth** [comment]
I'm a first time user of GdUnit3, and a long time writer of unit tests.  First up, let me say GdUnit3 is fantastic and is fulfilling my needs so far.  Please accept my heart-felt appreciation.  ❤️ 

Here's my setup:
* GdUnit3 `v2.3.0`
* Godot `v3.5.1.stable.official [6fed1ffa3]`
* Windows 11 22H2
* a fairly beefy Windows machine (Surface Studio i7 with 16Gb RAM)

Now for some feedback:

1. I have a performance issue.  I have a parameterised gdscript unit test with 49 items in the parameter array.  The test takes `24 seconds to start running` after I press "run unit tests" (or even "rerun unit tests") and about `825 milliseconds to execute` all 49 cases.
The long delay before starting is hard to take, especially when doing fast TDD cycles.
The test method takes 4 parameters (`String`, `Vector2`, `int`, `int`).
Any idea why there is such a long delay?  Any pointers on how I can diagnose?

2. When I right-click a parameterised test entry in the GdUnit3 tab and select either the `Debug Test` or `Run Test` option, after an eternity (i.e. the 24 second delay from feedback 1) the test runner returns immediately with no results.  Is this a bug?

Before:
![image](https://user-images.githubusercontent.com/546368/208291543-ddb8e21b-6284-4792-994f-d32b8c8ffab1.png)

After:
![image](https://user-images.githubusercontent.com/546368/208291591-e0255e24-5379-4fa3-837e-a58f4f0dd747.png)

3. When the tests finish running it collapses the tree in the GdUnit3 tab.  Can we please get an setting/option to leave the expand/collapse state of the tests as is before testing?

4. When there is a compile error in either the test function or the system under test (the code it is testing), the runner executes and goes green but nothing actually happens.  It's hard to tell that there is an error because of the way the tree collapses in feedback item 3.
  a. Can we not show the root test node as green in this case (so I know there is an issue)?
  b. Is it possible to show an error message?
I have to recognise there is an error (which is hard when the root node is green), debug the tests, to get the run-time error.

**Reply by MikeSchulze** [reply]
hi @ZodmanPerth thanks for your feedback.

To your questions.

1)
That sounds strange 24s only for test startup. Can you please create a issue and add an example to reproduce.
This will me help to find out what going wrong in your case.


2)
Execute a test via context menu form editor or from inspector should always have the same result.
I guess this bug is related to the first issue you described.

Execute a single step of an paramterized test is currently not supported, the context menu should be disabled at this selection.
If you want to execute a certain step of the paramterized test sounds like a valid use case.
Feel free to create feature request to imrove the usebility of GdUnit ;)

3)
By default a testsuite is collapsed if no errors are found. This will help to get a better overview over your failed tests to only let failing testsuites expanded.
Sure i can add a option to disable autocolapse succeeded testsuites.

4)
Compilation errors at runtime are not handled by Godot. They are simply ignored.
I recommend to always run the test in debug mode first, so that the debugger can show compilation errors.
To detect compilation errors at runtime, I implemented an error detection that must be enabled.
If this detection does not report your compile error as an error, please create a ticket with information and add steps to reproduce it.


![image](https://user-images.githubusercontent.com/347037/208294895-39fd732e-66b3-4118-9fd8-5ab587afd8dd.png)

**Reply by ZodmanPerth** [reply]
1. I have submitted a new issue #336

**Reply by ZodmanPerth** [reply]
2. I've created feature request  #337

**Reply by ZodmanPerth** [reply]
3. Thanks.  Should I make a feature request for that too?

**Reply by ZodmanPerth** [reply]
4.
> Compilation errors at runtime are not handled by Godot. They are simply ignored.

I see.  Thanks.

>To detect compilation errors at runtime, I implemented an error detection that must be enabled.
> If this detection does not report your compile error as an error, please create a ticket with information and add steps to reproduce it.

The errors are reported during a debug run.  Nothing more to do here.

**Reply by MikeSchulze** [reply]
Thank you, i will have a look and provide a fix.
Merry Christmas

**Reply by ZodmanPerth** [reply]
And to you!  🎁 🎅 🤶
